app/uk/gov/hmrc/emailverification/controllers/EmailPasscodeController.scala

```diff
@@ -0,0 +1,118 @@
+                                        passcodeRepo:PasscodeMongoRepository,
+                                        verifiedEmailRepo: VerifiedEmailMongoRepository,
+                                        analyticsConnector: PlatformAnalyticsConnector,
+                                        auditConnector: AuditConnector,
+                                        controllerComponents: ControllerComponents
+                                            )(implicit ec:ExecutionContext, appConfig: AppConfig) extends BaseControllerWithJsonErrorHandling(controllerComponents) {
+
+  def newPasscode(): String =  {
+    val codeSize = 6
+    Random.alphanumeric
+      .filterNot(_.isDigit)
+      .filterNot(_.isLower)
+      .filterNot(Set('A', 'E', 'I', 'O', 'U'))
+      .take(codeSize).mkString
+  }
+
+  private def sendEmailAndStorePasscode(request: PasscodeRequest)(implicit hc: HeaderCarrier) = {
+    val passcode = newPasscode()
+    val paramsWithPasscode = request.templateParameters.getOrElse(Map.empty) +
+      ("passcode" -> passcode)
+
+    for {
+      _ <- emailConnector.sendEmail(request.email, request.templateId, paramsWithPasscode)
+      _ <- passcodeRepo.upsert(request.sessionId, passcode, request.email, request.linkExpiryDuration)
+    } yield Created
+  }
+
+  def requestPasscode(): Action[JsValue] = Action.async(parse.json) {
+    implicit httpRequest =>
+      withJsonBody[PasscodeRequest] { request =>
+        verifiedEmailRepo.isVerified(request.email) flatMap {
+          case true => Future.successful(Conflict(Json.toJson(ErrorResponse("EMAIL_VERIFIED_ALREADY", "Email has already been verified"))))
+          case false =>
+            analyticsConnector.sendEvents(GaEvents.passcodeRequested)
+            sendEmailAndStorePasscode(request)
+        } recover {
+          case ex @ UpstreamErrorResponse(_, 400, _, _) =>
+            val event = ExtendedDataEvent(
+              auditSource =  "email-verification",
+              auditType = "AIV-60",
+              tags = hc.toAuditTags("requestPasscode", httpRequest.path),
+              detail = Json.obj(
+                "email-address" -> request.email,
+                "email-address-hex" -> toByteString(request.email)
+              )
+            )
+            auditConnector.sendExtendedEvent(event)
+            Logger.error("email-verification had a problem reading from repo", ex)
+            BadRequest(Json.toJson(ErrorResponse("BAD_EMAIL_REQUEST", ex.getMessage)))
+          case ex @ UpstreamErrorResponse(_, 404, _, _)  =>
+            Logger.error("email-verification had a problem, sendEmail returned not found", ex)
+            Status(BAD_GATEWAY)(Json.toJson(ErrorResponse("UPSTREAM_ERROR", ex.getMessage)))
+        }
+      }
+  }
+
+  private def toByteString(data: String) : String = {
+    data.getBytes("UTF-8").map("%02x".format(_)).mkString
+  }
+
+  private def storeEmailIfNotExist(email: String)(implicit hc: HeaderCarrier): Future[Result] = {
```

Comment:
opinion: this function seems so trivial, it might be better to inline it into `verifyPasscode`

Reply:
👍